src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,847 @@
+
+                    foreach (var platformSpecificOperation in platformSpecificOperations)
+                    {
+                        var value = analysisResult[platformSpecificOperation.Key.Kind, platformSpecificOperation.Key.Syntax];
+                        var attribute = platformSpecificOperation.Value;
+
+                        if (value.Kind == GlobalFlowStateAnalysisValueSetKind.Unknown)
+                        {
+                            if (platformSpecificOperation.Key.TryGetContainingLocalOrLambdaFunctionSymbol(out var containingSymbol))
+                            {
+                                var localResult = analysisResult.TryGetInterproceduralResultByDefinition(containingSymbol);
+                                if (localResult != null)
+                                {
+                                    var localValue = localResult[platformSpecificOperation.Key.Kind, platformSpecificOperation.Key.Syntax];
+                                    if (localValue.Kind == GlobalFlowStateAnalysisValueSetKind.Known && IsKnownValueGuarded(attribute, localValue))
+                                    {
+                                        continue;
+                                    }
+                                }
+                            }
+                        }
+                        else if (value.Kind == GlobalFlowStateAnalysisValueSetKind.Known && IsKnownValueGuarded(attribute, value))
+                        {
+                            continue;
+                        }
+
+                        ReportDiagnostics(platformSpecificOperation.Key, attribute, context);
+                    }
+                }
+                finally
+                {
+                    platformSpecificOperations.Free();
+                }
+
+                return;
+
+                OperationVisitor CreateOperationVisitor(GlobalFlowStateAnalysisContext context) => new OperationVisitor(guardMethods, osPlatformType, context);
+            });
+        }
+
+        private static bool IsKnownValueGuarded(SmallDictionary<string, PlatformAttributes> attributes, GlobalFlowStateAnalysisValueSet value)
+        {
+            if (value.AnalysisValues.Count == 1)
+            {
+                // No && or || operators used, result can be consumed directly
+                var analysisValue = value.AnalysisValues.First();
+
+                if (analysisValue is RuntimeMethodValue info &&
+                    attributes.TryGetValue(info.PlatformName, out var attribute))
+                {
+                    if (info.Negated)
+                    {
+                        if (attribute.UnsupportedFirst != null && IsEmptyVersion(attribute.UnsupportedFirst) && IsEmptyVersion(info.Version))
+                        {
+                            // the unsupported attribute suppressed setting null to not warn for it, same logic for all 
+                            attribute.UnsupportedFirst = null;
```

Comment:
Having publically mutable properties on `PlatformAttributes` makes me very uncomfortable. The fact that we mutate the the values of `SmallDictionary<string, PlatformAttributes> attributes` passed into the input makes me more worried - caller of `IsKnownValueGuarded(dictionary, ...)` would not expect the passed in dictionary to mutate during such a call. Would it be possible to avoid any mutation to inputs in this method?

1. This method should not be mutating any state on the input dictionary `SmallDictionary<string, PlatformAttributes> attributes`.

Reply:
Because of the requirement that API could have multiple platform attributes and which could be suppressed by different ways each attribute independently and end result depend on which ones suppressed which are not i had make it mutable. Even it is mutable attributes in the cache is not updated in the code, if there is an attribute need to be checked further we are creating [new attribute](https://github.com/dotnet/roslyn-analyzers/pull/3917/files#diff-ce2ca65813c46e221560ec2f25e5a89bR477) which used here and its safe to update this copy here as it is used only for current API